README.md
[![Build Status](https://travis-ci.com/worthmine/Text-vCard-Precisely.svg?branch=master)](https://travis-ci.com/worthmine/Text-vCard-Precisely) [![MetaCPAN Release](https://badge.fury.io/pl/Text-vCard-Precisely.svg)](https://metacpan.org/release/Text-vCard-Precisely) [![Build Status](https://img.shields.io/appveyor/ci/worthmine/Text-vCard-Precisely/master.svg?logo=appveyor)](https://ci.appveyor.com/project/worthmine/Text-vCard-Precisely/branch/master)
# NAME

Text::vCard::Precisely - Read, Write and Edit the vCards 3.0 and/or 4.0 precisely

# SYNOPSIS

    use Text::vCard::Precisely;
    my $vc = Text::vCard::Precisely->new();
    # Or now you can write like below if you want to use 4.0:
    my $vc4 = Text::vCard::Precisely->new( version => '4.0' );
    #or $vc4 = Text::vCard::Precisely::V4->new(); # it's same

    $vc->n([ 'Gump', 'Forrest', , 'Mr', '' ]);
    $vc->fn( 'Forrest Gump' );

    use GD;
    use MIME::Base64;
    my $gd = GD::Image->new( 100, 100 );
    my $black = $gd->colorAllocate( 0, 0, 0 );
    $gd->rectangle( 0, 0, 99, 99, $black );

    my $img = $gd->png();
    my $base64 = MIME::Base64::encode($img);

    $vc->photo([
       { content => 'https://avatars2.githubusercontent.com/u/2944869?v=3&s=400',  media_type => 'image/jpeg' },
       { content => $img, media_type => 'image/png' }, # Now you can set a binary image directly
       { content => $base64, media_type => 'image/png' }, # Also accept the text encoded in Base64
    ]);

    $vc->org('Bubba Gump Shrimp Co.'); # Now you can set/get org!

    $vc->tel({ content => '+1-111-555-1212', types => ['work'], pref => 1 });

    $vc->email({ content => 'forrestgump@example.com', types => ['work'] });

    $vc->adr( {
       types => ['work'],
       pobox     => '109',
       extended  => 'Shrimp Bld.',
       street    => 'Waters Edge',
       city      => 'Baytown',
       region    => 'LA',
       post_code => '30314',
       country   => 'United States of America',
    });

    $vc->url({ content => 'https://twitter.com/worthmine', types => ['twitter'] }); # for URL param

    print $vc->as_string();

# DESCRIPTION

A vCard is a digital business card.
vCard and [Text::vFile::asData](https://metacpan.org/pod/Text%3A%3AvFile%3A%3AasData) provides an API for parsing vCards

This module is forked from [Text::vCard](https://metacpan.org/pod/Text%3A%3AvCard)
because some reason below:

- Text::vCard **doesn't provide** full methods based on [RFC2426](https://tools.ietf.org/html/rfc2426)
- Mac OS X and iOS can't parse vCard4.0 with UTF-8 precisely. they cause some Mojibake
- Android 4.4.x can't parse vCard4.0

To handle an address book with several vCard entries in it, start with
[Text::vFile::asData](https://metacpan.org/pod/Text%3A%3AvFile%3A%3AasData) and then come back to this module.

Note that the vCard RFC requires `VERSION` and `FN`.  This module does not check or warn yet if these conditions have not been met

# Constructors

## load\_hashref($HashRef)

Accepts a HashRef that looks like below:

    my $hashref = {
       N   => [ 'Gump', 'Forrest', '', 'Mr.', '' ],
       FN  => 'Forrest Gump',
       SORT_STRING => 'Forrest Gump',
       ORG => 'Bubba Gump Shrimp Co.',
       TITLE => 'Shrimp Man',
       PHOTO => { media_type => 'image/gif', content => 'http://www.example.com/dir_photos/my_photo.gif' },
       TEL => [
           { types => ['WORK','VOICE'], content => '(111) 555-1212' },
           { types => ['HOME','VOICE'], content => '(404) 555-1212' },
       ],
       ADR =>[{
           types       => ['work'],
           pref        => 1,
           extended    => 100,
           street      => 'Waters Edge',
           city        => 'Baytown',
           region      => 'LA',
           post_code   => '30314',
           country     => 'United States of America'
       },{
           types       => ['home'],
           extended    => 42,
           street      => 'Plantation St.',
           city        => 'Baytown',
           region      => 'LA',
           post_code   => '30314',
           country     => 'United States of America'
       }],
       URL => 'http://www.example.com/dir_photos/my_photo.gif',
       EMAIL => 'forrestgump@example.com',
       REV => '2008-04-24T19:52:43Z',
    };

## load\_file($file\_name)

Accepts a file name

## load\_string($vCard)

Accepts a vCard string

# METHODS

## as\_string()

Returns the vCard as a string.
You have to use `Encode::encode_utf8()` if your vCard is written in utf8

## as\_file($filename)

Write data in vCard format to $filename.
Dies if not successful

# SIMPLE GETTERS/SETTERS

These methods accept and return strings

## version()

returns Version number of the vcard.
Defaults to **'3.0'** and this method is **READONLY**

## rev()

To specify revision information about the current vCard

## sort\_string()

To specify the family name, given name or organization text to be used for
national-language-specific sorting of the `FN`, `N` and `ORG`.

**This method is DEPRECATED in vCard4.0** Use `SORT-AS` param instead of it.

# COMPLEX GETTERS/SETTERS

They are based on Moose with coercion.
So these methods accept not only ArrayRef\[HashRef\] but also ArrayRef\[Str\],
single HashRef or single Str.

Read source if you were confused

## n()

To specify the components of the name of the object the vCard represents

## tel()

Accepts/returns an ArrayRef that looks like:

    [
       { types => ['work'], content => '651-290-1234', preferred => 1 },
       { types => ['home'], content => '651-290-1111' },
    ]

After version 0.18, **content will not be validated as phone numbers** 
All _Str_ types are accepted.

So you have to validate phone numbers with your way.

## adr(), address()

Accepts/returns an ArrayRef that looks like:

    [
       { types => ['work'], street => 'Main St', pref => 1 },
       {   types     => ['home'],
           pobox     => 1234,
           extended  => 'asdf',
           street    => 'Army St',
           city      => 'Desert Base',
           region    => '',
           post_code => '',
           country   => 'USA',
           pref      => 2,
       },
    ]

## email()

Accepts/returns an ArrayRef that looks like:

    [
       { types => ['work'], content => 'bbanner@ssh.secret.army.mil' },
       { types => ['home'], content => 'bbanner@timewarner.com', pref => 1 },
    ]

or accept the string as email like below

    'bbanner@timewarner.com'

## url()

Accepts/returns an ArrayRef that looks like:

    [
       { content => 'https://twitter.com/worthmine', types => ['twitter'] },
       { content => 'https://github.com/worthmine' },
    ]

or accept the string as URL like below

    'https://github.com/worthmine'

## photo(), logo()

Accepts/returns an ArrayRef of URLs or Images: 
Even if they are raw image binary or text encoded in Base64, it does not matter

**Attention!** Mac OS X and iOS **ignore** the description beeing URL

use Base64 encoding or raw image binary if you have to show the image you want

## note()

To specify supplemental information or a comment that is associated with the vCard

## org(), title(), role(), categories()

To specify additional information for your jobs

In these, `CATEGORIES` may have multiple content with being separated by COMMA.
multiple content is expressed by using ArrayRef like this:

    $vc->categories([qw(Internet Travel)]);

## fn(), full\_name(), fullname()

A person's entire name as they would like to see it displayed

## nickname()

To specify the text corresponding to the nickname of the object the vCard represents

Like `CATEGORIES`, It ALSO may have multiple content with being separated by COMMA.

    $vc->nickname([qw(Johny John)]);

## geo()

To specify information related to the global positioning of the object the vCard represents

## key()

To specify a public key or authentication certificate associated with the object that the vCard represents

## label()

ToDo: because **It's DEPRECATED in vCard4.0**

To specify the formatted text corresponding to delivery address of the object the vCard represents

## uid()

To specify a value that represents a globally unique identifier corresponding to 
the individual or resource associated with the vCard

## tz(), timezone()

Both are same method with Alias

To specify information related to the time zone of the object the vCard represents

utc-offset format is NOT RECOMMENDED from vCard4.0

`TZ` can be a URL, but there is no document in
 [RFC2426](https://tools.ietf.org/html/rfc2426)
or [RFC6350](https://tools.ietf.org/html/rfc6350)

So it just supports some text values

## bday(), birthday()

Both are same method with Alias

To specify the birth date of the object the vCard represents

## prodid()

To specify the identifier for the product that created the vCard object

## source()

To identify the source of directory information contained in the content type

## sound()

To specify a digital sound content information that annotates some aspect of the vCard

This property is often used to specify the proper pronunciation of 
the name property value of the vCard

## socialprofile()

There is no documents about `X-SOCIALPROFILE` in RFC but it works in iOS and Mac OS X!

I don't know well about in Android or Windows. Somebody please feedback me

## label()

**It's DEPRECATED in vCard4.0** You can use this method Just ONLY in vCard3.0

# For operating files with multiple vCards

See [Text::vCard::Precisely::Multiple](https://metacpan.org/pod/Text%3A%3AvCard%3A%3APrecisely%3A%3AMultiple)

# aroud UTF-8

If you want to send precisely the vCard with UTF-8 characters to the **ALMOST**
of smartphones, Use 3.0

It seems to be TOO EARLY to use 4.0

# for under perl-5.12.5

This module uses `\P{ascii}` in regexp so You have to use 5.12.5 and later

# SEE ALSO

- [RFC 2426](https://tools.ietf.org/html/rfc2426)
- [RFC 2425](https://tools.ietf.org/html/rfc2425)
- [RFC 6350](https://tools.ietf.org/html/rfc6350)
- [Text::vCard::Precisely::Multiple](https://metacpan.org/pod/Text%3A%3AvCard%3A%3APrecisely%3A%3AMultiple)
- [Text::vFile::asData](https://metacpan.org/pod/Text%3A%3AvFile%3A%3AasData)

# AUTHOR

Yuki Yoshida([worthmine](https://github.com/worthmine))

# LICENSE

This is free software; you can redistribute it and/or modify it under the same terms as Perl.


Changes
Revision history for Perl extension Text-vCard-Precisely

{{$NEXT}}

0.28 2020-10-22T14:12:40Z
  - started to accept newlines in some nodes like addresses

0.27 2020-10-16T12:55:18Z
  - correct spec for RFC2426 again
  - so categories and nikname can have multiple values separated by comma

0.26 2020-10-15T11:42:59Z
  - made document clear
  - make code & tests strict for RFC6350

0.25 2020-10-15T07:59:12Z
  - fixed #20 (url with hashref was broken)
  - renamed the tests for contexts

0.24 2020-10-15T01:56:14Z
  - bug fixed for not overridden with using 'Text::vCard::Precisely->new(version => '4.0');'
  - renamed the tests for correct order

0.23 2020-08-01T05:25:01Z
  - optimized the PODs for reading

0.22 2020-08-01T02:03:50Z
  - just fixed broken POD

0.21 2020-08-01T01:34:02Z
  - Added YouTube and Twitch as SocialProfile options
  - replaced wrong regexps
  - modified the PODs

0.20 2020-02-10T21:37:42Z
  - accomplished to fix the test for Windows
  - remove Facebook::Graph from SYNOPSIS

0.19 2020-02-08T12:13:03Z
  - multiple vCards into single file is now available
  - test on Windows will fail again

0.18 2020-02-02T12:46:21Z
  - gave up to validate phone numbers

0.17 2020-01-22T07:11:05Z
  - adupted with some environment they have no Text::Diff

0.16 2020-01-20T11:49:06Z
  - failing to install for Windows is fixed

0.15 2020-01-18T14:59:11Z
  - devided the test for as_file() with Windows or not  

0.14 2020-01-17T08:58:25Z
  - added the test and script for method as_file()

0.13 2017-10-21T12:03:26Z
  - just fixed around POD

0.12 2017-10-20T22:16:51Z
  - added supporting Instagram in X-SOCIALPROFILE 

0.11 2017-10-17T10:42:20Z
  - just optimized around coercing

0.10 2017-08-21T09:20:53Z
  - Now available SORT-AS param in N,FN,ORG

0.09 2017-08-18T11:25:47Z
  - fixed the tests for centOS

0.08 2017-08-18T10:35:49Z
  - fixed the tests for some environments

0.07 2017-08-18T02:20:29Z
  - fixed the tests for some environments

0.06 2017-08-17T13:33:23Z
  - migrate Dizil to Minilla

0.05 2017-08-11T11:56:13Z
  - fixed some tests

0.04 2017-08-01T15:19:21Z
  - first cpan publishment

0.03 2017-06-07T07:02:46Z
  - be more strictly

0.02 2017-05-23T12:43:30Z
  - original version


lib/Text/vCard/Precisely.pm
package Text::vCard::Precisely;

our $VERSION = '0.28';

use Moose;
use Moose::Util::TypeConstraints;

extends 'Text::vCard::Precisely::V3';

enum 'Version' => [qw( 3.0 4.0 )];
has version    => ( is => 'ro', isa => 'Version', default => '3.0', required => 1 );

sub BUILD {
    my $self = shift;
    return Text::vCard::Precisely::V3->new(@_) unless $self->version() eq '4.0';

    require Text::vCard::Precisely::V4;
    our @ISA = 'Text::vCard::Precisely::V4';
    return Text::vCard::Precisely::V4->new(@_);
}

__PACKAGE__->meta->make_immutable;
no Moose;

1;

__END__

=encoding UTF8

=head1 NAME

Text::vCard::Precisely - Read, Write and Edit the vCards 3.0 and/or 4.0 precisely

=head1 SYNOPSIS
 
 use Text::vCard::Precisely;
 my $vc = Text::vCard::Precisely->new();
 # Or now you can write like below if you want to use 4.0:
 my $vc4 = Text::vCard::Precisely->new( version => '4.0' );
 #or $vc4 = Text::vCard::Precisely::V4->new(); # it's same

 $vc->n([ 'Gump', 'Forrest', , 'Mr', '' ]);
 $vc->fn( 'Forrest Gump' );

 use GD;
 use MIME::Base64;
 my $gd = GD::Image->new( 100, 100 );
 my $black = $gd->colorAllocate( 0, 0, 0 );
 $gd->rectangle( 0, 0, 99, 99, $black );

 my $img = $gd->png();
 my $base64 = MIME::Base64::encode($img);

 $vc->photo([
    { content => 'https://avatars2.githubusercontent.com/u/2944869?v=3&s=400',  media_type => 'image/jpeg' },
    { content => $img, media_type => 'image/png' }, # Now you can set a binary image directly
    { content => $base64, media_type => 'image/png' }, # Also accept the text encoded in Base64
 ]);

 $vc->org('Bubba Gump Shrimp Co.'); # Now you can set/get org!

 $vc->tel({ content => '+1-111-555-1212', types => ['work'], pref => 1 });

 $vc->email({ content => 'forrestgump@example.com', types => ['work'] });

 $vc->adr( {
    types => ['work'],
    pobox     => '109',
    extended  => 'Shrimp Bld.',
    street    => 'Waters Edge',
    city      => 'Baytown',
    region    => 'LA',
    post_code => '30314',
    country   => 'United States of America',
 });

 $vc->url({ content => 'https://twitter.com/worthmine', types => ['twitter'] }); # for URL param

 print $vc->as_string();

=head1 DESCRIPTION

A vCard is a digital business card.
vCard and L<Text::vFile::asData> provides an API for parsing vCards

This module is forked from L<Text::vCard>
because some reason below:

=over

=item

Text::vCard B<doesn't provide> full methods based on L<RFC2426|https://tools.ietf.org/html/rfc2426>

=item

Mac OS X and iOS can't parse vCard4.0 with UTF-8 precisely. they cause some Mojibake

=item

Android 4.4.x can't parse vCard4.0

=back

To handle an address book with several vCard entries in it, start with
L<Text::vFile::asData> and then come back to this module.

Note that the vCard RFC requires C<VERSION> and C<FN>.  This module does not check or warn yet if these conditions have not been met

=head1 Constructors

=head2 load_hashref($HashRef)

Accepts a HashRef that looks like below:

 my $hashref = {
    N   => [ 'Gump', 'Forrest', '', 'Mr.', '' ],
    FN  => 'Forrest Gump',
    SORT_STRING => 'Forrest Gump',
    ORG => 'Bubba Gump Shrimp Co.',
    TITLE => 'Shrimp Man',
    PHOTO => { media_type => 'image/gif', content => 'http://www.example.com/dir_photos/my_photo.gif' },
    TEL => [
        { types => ['WORK','VOICE'], content => '(111) 555-1212' },
        { types => ['HOME','VOICE'], content => '(404) 555-1212' },
    ],
    ADR =>[{
        types       => ['work'],
        pref        => 1,
        extended    => 100,
        street      => 'Waters Edge',
        city        => 'Baytown',
        region      => 'LA',
        post_code   => '30314',
        country     => 'United States of America'
    },{
        types       => ['home'],
        extended    => 42,
        street      => 'Plantation St.',
        city        => 'Baytown',
        region      => 'LA',
        post_code   => '30314',
        country     => 'United States of America'
    }],
    URL => 'http://www.example.com/dir_photos/my_photo.gif',
    EMAIL => 'forrestgump@example.com',
    REV => '2008-04-24T19:52:43Z',
 };

=head2 load_file($file_name)

Accepts a file name

=head2 load_string($vCard)

Accepts a vCard string

=head1 METHODS

=head2 as_string()

Returns the vCard as a string.
You have to use C<Encode::encode_utf8()> if your vCard is written in utf8

=head2 as_file($filename)

Write data in vCard format to $filename.
Dies if not successful

=head1 SIMPLE GETTERS/SETTERS

These methods accept and return strings

=head2 version()

returns Version number of the vcard.
Defaults to B<'3.0'> and this method is B<READONLY>

=head2 rev()

To specify revision information about the current vCard

=head2 sort_string()

To specify the family name, given name or organization text to be used for
national-language-specific sorting of the C<FN>, C<N> and C<ORG>.

B<This method is DEPRECATED in vCard4.0> Use C<SORT-AS> param instead of it.

=head1 COMPLEX GETTERS/SETTERS

They are based on Moose with coercion.
So these methods accept not only ArrayRef[HashRef] but also ArrayRef[Str],
single HashRef or single Str.

Read source if you were confused

=head2 n()

To specify the components of the name of the object the vCard represents

=head2 tel()

Accepts/returns an ArrayRef that looks like:

 [
    { types => ['work'], content => '651-290-1234', preferred => 1 },
    { types => ['home'], content => '651-290-1111' },
 ]

After version 0.18, B<content will not be validated as phone numbers> 
All I<Str> types are accepted.

So you have to validate phone numbers with your way.

=head2 adr(), address()

Accepts/returns an ArrayRef that looks like:

 [
    { types => ['work'], street => 'Main St', pref => 1 },
    {   types     => ['home'],
        pobox     => 1234,
        extended  => 'asdf',
        street    => 'Army St',
        city      => 'Desert Base',
        region    => '',
        post_code => '',
        country   => 'USA',
        pref      => 2,
    },
 ]

=head2 email()

Accepts/returns an ArrayRef that looks like:

 [
    { types => ['work'], content => 'bbanner@ssh.secret.army.mil' },
    { types => ['home'], content => 'bbanner@timewarner.com', pref => 1 },
 ]

or accept the string as email like below

 'bbanner@timewarner.com'

=head2 url()

Accepts/returns an ArrayRef that looks like:

 [
    { content => 'https://twitter.com/worthmine', types => ['twitter'] },
    { content => 'https://github.com/worthmine' },
 ]

or accept the string as URL like below

 'https://github.com/worthmine'

=head2 photo(), logo()

Accepts/returns an ArrayRef of URLs or Images: 
Even if they are raw image binary or text encoded in Base64, it does not matter

B<Attention!> Mac OS X and iOS B<ignore> the description beeing URL

use Base64 encoding or raw image binary if you have to show the image you want

=head2 note()

To specify supplemental information or a comment that is associated with the vCard

=head2 org(), title(), role(), categories()

To specify additional information for your jobs

In these, C<CATEGORIES> may have multiple content with being separated by COMMA.
multiple content is expressed by using ArrayRef like this:

 $vc->categories([qw(Internet Travel)]);

=head2 fn(), full_name(), fullname()

A person's entire name as they would like to see it displayed

=head2 nickname()

To specify the text corresponding to the nickname of the object the vCard represents

Like C<CATEGORIES>, It ALSO may have multiple content with being separated by COMMA.

 $vc->nickname([qw(Johny John)]);

=head2 geo()

To specify information related to the global positioning of the object the vCard represents

=head2 key()

To specify a public key or authentication certificate associated with the object that the vCard represents

=head2 label()

ToDo: because B<It's DEPRECATED in vCard4.0>

To specify the formatted text corresponding to delivery address of the object the vCard represents

=head2 uid()

To specify a value that represents a globally unique identifier corresponding to 
the individual or resource associated with the vCard

=head2 tz(), timezone()

Both are same method with Alias

To specify information related to the time zone of the object the vCard represents

utc-offset format is NOT RECOMMENDED from vCard4.0

C<TZ> can be a URL, but there is no document in
 L<RFC2426|https://tools.ietf.org/html/rfc2426>
or L<RFC6350|https://tools.ietf.org/html/rfc6350>

So it just supports some text values

=head2 bday(), birthday()

Both are same method with Alias

To specify the birth date of the object the vCard represents

=head2 prodid()

To specify the identifier for the product that created the vCard object

=head2 source()

To identify the source of directory information contained in the content type

=head2 sound()

To specify a digital sound content information that annotates some aspect of the vCard

This property is often used to specify the proper pronunciation of 
the name property value of the vCard

=head2 socialprofile()

There is no documents about C<X-SOCIALPROFILE> in RFC but it works in iOS and Mac OS X!

I don't know well about in Android or Windows. Somebody please feedback me

=head2 label()

B<It's DEPRECATED in vCard4.0> You can use this method Just ONLY in vCard3.0

=head1 For operating files with multiple vCards

See L<Text::vCard::Precisely::Multiple>

=head1 aroud UTF-8

If you want to send precisely the vCard with UTF-8 characters to the B<ALMOST>
of smartphones, Use 3.0

It seems to be TOO EARLY to use 4.0

=head1 for under perl-5.12.5

This module uses C<\P{ascii}> in regexp so You have to use 5.12.5 and later

=head1 SEE ALSO

=over

=item

L<RFC 2426|https://tools.ietf.org/html/rfc2426>

=item

L<RFC 2425|https://tools.ietf.org/html/rfc2425>

=item

L<RFC 6350|https://tools.ietf.org/html/rfc6350>

=item

L<Text::vCard::Precisely::Multiple>

=item

L<Text::vFile::asData>

=back

=head1 AUTHOR

Yuki Yoshida(L<worthmine|https://github.com/worthmine>)

=head1 LICENSE

This is free software; you can redistribute it and/or modify it under the same terms as Perl.
